# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: accessible/generic/DocAccessible.cpp
# Commit: 6591f14bbc71
# Full Hash: 6591f14bbc713f11b6316712d931ecc19a443a5c
# Author: Morgan Reschenberg <mreschenberg@mozilla.com>
# Date: 2021-10-05 03:52:32
# Regressor Bug: 1726227
# File Overlap Count: 1
# Description:
#   Bug 1726227: Cache parent-relative accessible bounds, resolution in parent process r=Jamie,eeejay
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D123399
# ==============================================================================

diff -r 06dda9072cb7 -r 6591f14bbc71 accessible/generic/DocAccessible.cpp
--- a/accessible/generic/DocAccessible.cpp	Mon Oct 04 19:28:49 2021 +0000
+++ b/accessible/generic/DocAccessible.cpp	Mon Oct 04 19:46:54 2021 +0000
@@ -908,6 +908,10 @@
   ContentRemoved(aChildNode);
 }
 
+void DocAccessible::MarkForBoundsProcessing(LocalAccessible* aAcc) {
+  mMaybeBoundsChanged.EnsureInserted(aAcc);
+}
+
 void DocAccessible::ParentChainChanged(nsIContent* aContent) {}
 
 ////////////////////////////////////////////////////////////////////////////////
@@ -1335,6 +1339,31 @@
   mInvalidationList.Clear();
 }
 
+void DocAccessible::ProcessBoundsChanged() {
+  if (!StaticPrefs::accessibility_cache_enabled_AtStartup()) {
+    return;
+  }
+
+  nsTArray<CacheData> data;
+  for (auto acc : mMaybeBoundsChanged) {
+    if (!acc->IsDefunct()) {
+      RefPtr<AccAttributes> fields = acc->BundleFieldsForCache(
+          CacheDomain::Bounds, CacheUpdateType::Update);
+      if (fields->Count()) {
+        data.AppendElement(CacheData(
+            acc->IsDoc() ? 0 : reinterpret_cast<uint64_t>(acc->UniqueID()),
+            fields));
+      }
+    }
+  }
+
+  mMaybeBoundsChanged.Clear();
+
+  if (data.Length()) {
+    IPCDoc()->SendCache(CacheUpdateType::Update, data, true);
+  }
+}
+
 LocalAccessible* DocAccessible::GetAccessibleEvenIfNotInMap(
     nsINode* aNode) const {
   if (!aNode->IsContent() ||